### PR TITLE
feature[design]: allow design plugin to interface with pytorch via Result

### PR DESCRIPTION
**Pull request by groberts-flex**
This PR fixes a small bug in the design plugin wherein the optimization method returns the dimensions in a different order and so labeling with the `self.dims` parameter can end up being incorrect.

It also implements accessor and length functions for the Result class so that it can be used directly as a Dataset in PyTorch. This also may be useful generally.

Marking this as a draft for now to discuss if we want to make these changes to the plugin. The changes are currently needed for the surrogate notebooks draft PR. (https://github.com/flexcompute/tidy3d-notebooks/pull/228)

### CODE REVIEW

**Comment by groberts-flex** [comment]
converting now to a PR since the notebook PR is out

File: tidy3d/plugins/design/design.py [hunk 1]
```diff
@@ -124,7 +124,7 @@ def _package_run_results(
         fn_args_coords_T = list(map(list, zip(*fn_args_coords)))
 
         return Result(
-            dims=self.dims,
+            dims=tuple(fn_args[0].keys()),
```
Comment (yaugenst-flex) [inline tidy3d/plugins/design/design.py:127]: Not too familiar with the design plugin, why is this change necessary?

File: tidy3d/plugins/design/result.py [hunk 2]
```diff
@@ -427,3 +427,18 @@ def add(self, fn_args: Dict[str, float], value: Any) -> Result:
...
+    def __len__(self):
+        return len(self.coords)
+
+    def __getitem__(self, data_index):
+        features = self.coords[data_index]
+        labels = self.values[data_index]
+
+        return features, labels
+
+    def __getitems__(self, data_indices):
```
Comment (yaugenst-flex) [inline tidy3d/plugins/design/result.py]: Is this required? Since it's not a standard python getter, could this class also work without it? I assume pytorch would just automatically batch using `__getitem__` instead?

**Review by yaugenst-flex — COMMENTED** [review]
Thanks @groberts-flex the accessors definitely are a great addition!

File: tidy3d/plugins/design/design.py (hunk 1 above)
Comment (groberts-flex) [reply tidy3d/plugins/design/design.py:127]: Right now, the optimization methods don't necessarily return the values in the same order as they are in the DesignSpace dims attribute (which I believe is just set by the order the user inputs the parameters). I think the BayOpt for example returns with dimensions in alphabetical order. So self.dims is not always the correct key ordering for the result.

File: tidy3d/plugins/design/result.py (hunk 2 above)
Comment (groberts-flex) [reply tidy3d/plugins/design/result.py]: good point on this! I thought it needed to be implemented, but actually just __getitem__ needs to be implemented. I think the __getitems__ can be used in batch sampling but if not available I think it just implements it using the underlying __getitem__ which is basically what I did here. will test this with the notebook, but think it can be removed.

File: tidy3d/plugins/design/design.py (hunk 1 above)
Comment (yaugenst-flex) [reply tidy3d/plugins/design/design.py:127]: I see. Does this mean then that the `dims` of a `Result` don't necessarily align with the `dims` of the `DesignSpace` they were generated from? And if yes, is that intended behavior or should we try to align them?

Comment (groberts-flex) [reply tidy3d/plugins/design/design.py:127]: Yeah, they don't necessarily align (in fact they weren't aligning in one of the surrogate notebooks and it was causing the data to be interpreted backwards) because of the method outputting them in a different order.

I think it would be ideal if all the parameters stayed in the same order as originally specified (or always were shuffled to the same ordering). Right now, the Result is also packaged up with the optimizer that was used in the method. And a user can query that object to find the parameters associated with the optimal value. Right now, with the current change, the Result returned from here is consistent with the data and the optimizer. But if we rearrange here to match the DesignSpace, we should probably rearrange the optimizer parameters too, which I think is not straightforward since I believe that's coming from an external package.

Comment (yaugenst-flex) [reply tidy3d/plugins/design/design.py:127]: Alright, ok. Something to keep in mind then, but let's keep it as is for this PR then

File: tidy3d/plugins/design/result.py (hunk 2 above)
Comment (groberts-flex) [reply tidy3d/plugins/design/result.py]: confirmed that we don't need getitems! also returning these as numpy arrays works better with PyTorch integration (and possibly generally). they come out as tensors from the dataloader now instead of a list of tensors which was causing issues before.

**Comment by yaugenst-flex** [comment]
Thanks @groberts-flex I think this is good to go implementation-wise, but still needs tests and a changelog entry. For the tests, I think the accessor methods can be tested easily (check `len`, indexing, `list/tuple` conversion). For the dimension ordering, maybe identify one (or ideally more than one distinct) cases where this broke previously?

**Comment by groberts-flex** [comment]
tests and changelog entry are in, how does this look for merging @yaugenst-flex?

File: tests/test_plugins/test_design.py [hunk 3]
```diff
@@ -35,6 +35,28 @@
...
+    ),
+    dict(
+        dims=("x", "y", "z", "t"),
+        coords=((1, 2, 3, 4), (5, 6, 7, 8)),
+        values=(5, 3),
+        multi_val=False,
+    ),
+    dict(dims=tuple("a"), coords=(tuple("b"), tuple("c")), values=("d", "e"), multi_val=False),
+]
+
```
Comment (yaugenst-flex) [inline tests/test_plugins/test_design.py]: since this isn't reused for other tests and the cases are relatively simple, i'd prefer this to be written out in the parametrization directly, i.e.
```python
@pytest.mark.parametrize("dims, coords, values, multi_val", [
    (("a", "b", "c"), ((1, 0, 1), (2, 3, 4), (5, 1, 2)), (1, 2, 3), False),
    # ... other cases
])
def test_result_accessor(dims, coords, values, multi_val):
    # ... test logic
```

this puts the data closer to the test, making it easier to see the test cases when looking at the test, and the function arguments are clearer too

**Review by yaugenst-flex — APPROVED** [review]
Thanks @groberts-flex I think this looks good, only a minor comment.

File: tests/test_plugins/test_design.py (hunk 3 above)
Comment (groberts-flex) [reply tests/test_plugins/test_design.py]: thanks, this makes sense! much clearer this way
